src/perl.c

```diff
@@ -2117,12 +2159,30 @@ static int perl_shutdown (void)
 	t = perl_threads->tail;
 
 	while (NULL != t) {
+		struct timespec ts_wait;
 		c_ithread_t *thr = t;
 
 		/* the pointer has to be advanced before destroying
 		 * the thread as this will free the memory */
 		t = t->prev;
 
+		thr->shutdown = 1;
+		if (thr->running) {
+			/* Give some time to thread to exit from pi */
+			WARNING ("perl shutdown: thread is running inside perl. Waiting.");
+			ts_wait.tv_sec = 0;
+			ts_wait.tv_nsec = 500000;
+			nanosleep (&ts_wait, NULL);
```

Comment:
I have no idea how to stop program execution more correctly than proposed solution. 
This code handles perl interpreter on threads which was started by other Collectd plugins also. 
Perl interpreter can be attached to such threads too, like to other threads.

Plugins should shutdown their threads by himself and perl plugin can't (and should not) affect this. 
The perl plugin can only finish his work (if running) and deny future calls to perl interpreter.

If perl plugin code works correctly, it will do his work fast and program (thread) execution will return to collectd/plugin.  No new enter to perl interpreter will happen due to 'shutdown' flag, attached to thread.

If perl plugin code does not finish his work, I found no way to terminate it gracefully.
I even have no idea how that can be implemented. 
SIGTERM is simply an attempt to terminate a separate thread to avoid process crash. 
That is only attempt without any certainty of success.

From my point of view, we give all required to perl plugin developers to make correct implementations.
If perl plugin implementation will be correct (and even if not quite correct), all things will happen successfully.